cpp/include/raft/linalg/map_then_reduce.cuh

```diff
@@ -82,9 +86,65 @@ void mapThenReduce(OutType* out,
                    const InType* in,
                    Args... args)
 {
-  detail::mapThenReduceImpl<InType, OutType, MapOp, ReduceLambda, TPB, Args...>(
+  detail::mapThenReduceImpl<InType, OutType, IdxType, MapOp, ReduceLambda, TPB, Args...>(
     out, len, neutral, map, op, stream, in, args...);
 }
+
+/**
+ * @defgroup map_reduce Map-Reduce ops
+ * @{
+ */
+
+/**
+ * @brief CUDA version of map and then generic reduction operation
+ * @tparam InElementType the data-type of the input
+ * @tparam OutElementType the data-type of the output
+ * @tparam MapOp the device-lambda performing the actual map operation
+ * @tparam ReduceLambda the device-lambda performing the actual reduction
+ * @tparam IndexType the index type
+ * @tparam TPB threads-per-block in the final kernel launched
+ * @tparam Args additional parameters
+ * @param handle raft::handle_t
+ * @param out the output reduced value assumed to be a raft::device_scalar_view
+ * @param neutral The neutral element of the reduction operation. For example:
+ *    0 for sum, 1 for multiply, +Inf for Min, -Inf for Max
+ * @param map the device-lambda
+ * @param op the reduction device lambda
+ * @param in the input of type raft::device_vector_view
+ * @param args additional input arrays
+ */
+template <typename InElementType,
+          typename MapOp,
+          typename ReduceLambda,
+          typename IndexType      = std::uint32_t,
+          typename OutElementType = InElementType,
+          int TPB                 = 256,
+          typename... Args>
+void map_reduce(const raft::handle_t& handle,
+                raft::device_scalar_view<OutElementType> out,
```

Comment:
We need to keep the order of arguments consistent across the new API (which means not following the order of the arguments from the existing API functions in many cases). I captured the expected orderings [here](https://github.com/rapidsai/raft/issues/535). Following a more thrust-like ordering, we should put the handle first, then inputs, then outputs, then additional params/arguments.